RandLAPACK/drivers/rl_revd2.hh

```diff
@@ -0,0 +1,240 @@
+        /// This is motivated by the fact that the approx error will never be smaller than nu.
+        /// This code is identical to ALgorithm E2 from https://arxiv.org/pdf/2110.02820.pdf.
+        /// Uses RangeFinder for constructing a sketching operator.
+        /// Has a lot of potential in terms of storage space optimization, 
+        /// which, however, will affect readability.
+        ///
+        /// @param[in] m
+        ///     The number of rows in the matrix A.
+        ///
+        /// @param[in] n
+        ///     The number of columns in the matrix A.
+        ///
+        /// @param[in] A
+        ///     The m-by-m matrix A, stored in a column-major format.
+        ///     Must be SPD.
+        ///
+        /// @param[in] k
+        ///     Column dimension of a sketch, k <= n.
+        ///
+        /// @param[in] V
+        ///     On entry, is empty and may not have any space allocated for it.
+        ///
+        /// @param[in] E
+        ///     On entry, is empty and may not have any space allocated for it.
+        ///
+        /// @param[out] V
+        ///     Stores m-by-k matrix matrix of eigenvectors.
+        ///
+        /// @param[out] E
+        ///     Stores k eigenvalues.
+        ///
+        /// @return = 0: successful exit
+        ///
+
+        int call(
+            int64_t m,
+            std::vector<T>& A,
+            int64_t& k,
+            std::vector<T>& V,
+            std::vector<T>& E
+        ) override;
+
+    public:
+        RandLAPACK::RangeFinder<T>& RF_Obj;
+        RandBLAS::base::RNGState<r123::Philox4x32> state;
+        T tol;
+        int p;
+        bool verbosity;
+
+        std::vector<T> Y;
+        std::vector<T> Omega;
+
+        std::vector<T> R;
+        std::vector<T> S;
+
+        std::vector<T> A_cpy;
+};
+
+// -----------------------------------------------------------------------------
+template <typename T>
+int REVD2<T>::call(
+        int64_t m, // m = n
+        std::vector<T>& A,
+        int64_t& k,
+        std::vector<T>& V,
+        std::vector<T>& E
+){
+    T err = 0;
+    while(1) {
+        T* A_dat = A.data();
+        T* V_dat = util::upsize(m * k, V);
+        util::upsize(k, E);
+        T* Y_dat = util::upsize(m * k, this->Y);
+        T* Omega_dat = util::upsize(m * k, this->Omega);
+        T* R_dat = util::upsize(k * k, this->R);
+        T* S_dat = util::upsize(k * k, this->S);
+
+        // Construnct a sketching operator
+        // If CholeskyQR is used for stab/orth here, RF fails
+        if(this->RF_Obj.call(m, m, A, k, this->Omega)) {
+            printf("RF/RS FAILED\n");
+            return 2;
+        }
+
+        // Y = A * S
+        blas::gemm(Layout::ColMajor, Op::NoTrans, Op::NoTrans, m, k, m, 1.0, A_dat, m, Omega_dat, m, 0.0, Y_dat, m);
+
+        T nu = std::sqrt(m) * std::numeric_limits<double>::epsilon() * lapack::lange(lapack::Norm::Fro, m, k, Y_dat, m);
```

Comment:
Do we really have the ``sqrt(m)`` here when using the Frobenius norm? When I see a ``sqrt(m)`` I think of the worst-case factor by which the Frobenius norm can over-estimate the spectral norm.

Reply:
You're right, we can drop the sqrt(m) in Fro norm case.

Reply:
Actually, we begin to overestimate rank really badly when we drop sqrt.

Reply:
So if we drop sqrt(m), nu is 7.389831e-16 for our test cases, which is never achieved by the algorithm, hence overestimation is happening.

Reply:
Anyway, this modification makes sense. I'll just set tol to 10^-14 here